### ISSUE TITLE
migrate to libunwind

### ISSUE THREAD

**Issue by gusev-p** [scylladb/seastar#1878]
Currently Seastar uses `glibc` `backtrace` function for collecting backtraces, it's not async-signal-safe and causes ASAN failures or ARM [issue1](https://github.com/scylladb/scylladb/issues/15090#issuecomment-1763355796), [issue2](https://github.com/scylladb/scylladb/issues/15090#issuecomment-1763355796). I have a stable reproducer script for these problems, it consistently fails with glibc backtrace and works with libunwind.

Here's what the `glibc` [manual says](https://github.com/bminor/glibc/blob/69239bd7a216007692470aa9d5f3658024638742/manual/debug.texi#L39) about `backtrace` async safety:
> @c The generic implementation just does pointer chasing within the local
> @c stack, without any guarantees that this will handle signal frames
> @c correctly, so it's AS-Unsafe to begin with.  However, most (all?)
> @c arches defer to libgcc_s's _Unwind_* implementation, dlopening
> @c libgcc_s.so to that end except in a static version of libc.
> @c libgcc_s's implementation may in turn defer to libunwind.  We can't
> @c assume those implementations are AS- or AC-safe, but even if we
> @c could, our own initialization path isn't, and libgcc's implementation
> @c calls malloc and performs internal locking, so...

[Here](https://blog.openj9.org/2022/03/16/collecting-native-stack-traces-on-macos-with-libunwind/) there is the following reasoning:
> libunwind is a library that is included in clang and available in MacOS 10.6 and later. It is an efficient C API that allows determining call stack information of the currently executing thread. This method of backtracing was chosen instead of the BSD backtrace because it is thread and signal safe, and provides more flexibility when handling signal-handler frames – MacOS modifies the stack during a call to a signal handler function.

It's about MacOS, but I guess it's applicable to `glibc` on linux as well. Here's [another claim](https://eli.thegreenplace.net/2015/programmatic-access-to-the-call-stack-in-c/) that 

> libunwind is the most modern, widespread and portable solution.

[Here](https://stackoverflow.com/questions/75944183/is-llvm-libunwind-safe-to-call-from-signal-handler) it's explicitly claimed that libunwind (nongnu, llvm implementation should be similar in this regard) is signal safe:
> The manual page for each libunwind routine identifies whether or not it is signal-safe, but as a general rule, any routine that may be needed for local unwinding is signal-safe (e.g., unw_step() for local unwinding is signal-safe). For remote-unwinding, none of the libunwind routines are guaranteed to be signal-safe.

Scylla used libunwind in the past, but in [this commit](https://github.com/scylladb/seastar/commit/a48fe69724492ffec64625a081f6dea54a50e470?diff=split), it was changed to glibc backtrace, we should probably revert this commit. I tried this locally and it worked fine. (for some reason `unw_backtrace` returns only the last frame, but in this code we iterate via `unw_step` and it works).

**Comment by nyh** [comment]
>. I have a stable reproducer script for these problems, it consistently fails with glibc backtrace and works with libunwind.

Nice, can you turn it into a unit test, so we don't regress this again when somebody decides to switch the backtrace implementation yet again?

> Scylla used libunwind in the past, but in [this commit](https://github.com/scylladb/seastar/commit/a48fe69724492ffec64625a081f6dea54a50e470?diff=split), it was changed to glibc backtrace, we should probably revert this commit. I tried this locally and it worked fine. (for some reason `unw_backtrace` returns only the last frame, but in this code we iterate via `unw_step` and it works).

Interesting. In that pull request, @avikivity said he switched away from libunwind because "libunwind is not available everywhere.". @avikivity do you remember why you said this, and where it was not available? If it's really not available in some architectures or OSs we want to support, then maybe we'll need to support both options? But maybe this 5-year-old statement is simply no longer true, and these days libunwind **is** available everywhere, and we can indeed just revert that patch.

**Comment by avikivity** [comment]
s390x, was fixed in https://src.fedoraproject.org/rpms/libunwind/c/5d7bc2445160ca914c927ed302d412f7b21906a0.

**Comment by gusev-p** [comment]
> Nice, can you turn it into a unit test, so we don't regress this again when somebody decides to switch the backtrace implementation yet again?

Yes, I though about that, I'll try to write such a test. It would be randomised, probably taking backtrace many times in a signal handler.

**Comment by gusev-p** [comment]
tried to run this on ARM/debug/ASAN - no crash:
```
SEASTAR_THREAD_TEST_CASE(test_backtrace) {
    std::atomic<unsigned> reports{};
    std::atomic<unsigned> frames{};
    temporary_stall_detector_settings tsds(10ms, [&]() {
        ++reports;
        backtrace([&](auto) {
            ++frames;
        });
    });
    auto body = [] { spin_user_hires(1ms); };
    constexpr unsigned nr = 500;
    for (unsigned i = 0; i < nr; ++i) {
        spin_some_cooperatively(100ms, body);
        spin(20ms, body);
    }
    BOOST_CHECK_EQUAL(reports, nr);
    BOOST_CHECK_GE(frames, 2 * nr);
}
```

My initial reproducer script ran scylla binary many times (~100). Since the crash happened in `make_scheduling_group_key_config<service::storage_proxy_stats::stats>()` in `scylla_main`,  I added `_exit(0)` after it to speedup iterations. Maybe we can write a similar test in python/test.py, inject `_exit(0)` through error injection mechanism. Not sure whether it is worth it though.

**Comment by Svelar** [comment]
Hi, any update so far?
